File: dotnet/src/internal.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Nito.AsyncEx;

public static class Program {
    public static readonly DateTime Start = DateTime.Now;
    static void Main(string[] args) {
        Example.Run();
    }
}

public class MessageBus: IMessageBus {
    readonly ConcurrentDictionary<string, AsyncProducerConsumerQueue<string>> queues = new ConcurrentDictionary<string, AsyncProducerConsumerQueue<string>>();

    public async Task Publish(string topic, string message) {
        await Publish(topic, message, CancellationToken.None);
    }

    public async Task Publish(string topic, string message, CancellationToken cancel) {
        await Task.Delay(10, cancel);
        var queue = this.queues.GetOrAdd(topic, _ => new AsyncProducerConsumerQueue<string>());
        await queue.EnqueueAsync(message, cancel);
    }

    public async Task<string> Receive(string topic) {
        return await Receive(topic, CancellationToken.None);
    }

    public async Task<string> Receive(string topic, CancellationToken cancel) {
        var queue = this.queues.GetOrAdd(topic, _ => new AsyncProducerConsumerQueue<string>());
        return await queue.DequeueAsync(cancel);
    }
}

public class Service {
    readonly IMessageBus bus;
    readonly Action<string> log = msg => Console.WriteLine($"{(DateTime.Now - Program.Start):s'.'fff} [Service] {msg}");
    readonly CancellationTokenSource Cancel = new CancellationTokenSource();

    public Service(IMessageBus bus) {
        this.bus = bus;
    }

    public async Task Run() {
        log($"[Run] Service starting");
        while (!this.Cancel.IsCancellationRequested) {
            try {
                var msg = await this.bus.Receive("/service", this.Cancel.Token);
                log($"[Run] Got start message: {msg}");
                await Operation(int.Parse(msg));
                log($"[Run] Handled msg");
            } catch (OperationCanceledException) {
            }
        }
        log($"[Run] Service stopping");
    }

    public void Stop() {
        this.Cancel.Cancel();
    }

    async Task Operation(int foo) {
        log($"[Operation] Starting operation");
        var op_cancel = new CancellationTokenSource();
        var operation = RunOperation(foo, op_cancel.Token);
        for (;;) {
            var msg_cancel = new CancellationTokenSource();
            var msg = this.bus.Receive("/service/op/cancel", msg_cancel.Token);
            var first = await Task.WhenAny(operation, msg);
            if (first == msg && await msg == "cancel") {
                log($"[Operation] Received cancel msg");
                op_cancel.Cancel();
                return;
            }
            msg_cancel.Cancel();
            if (first == operation) {
                log($"[Operation] finished");
                await this.bus.Publish($"/service/op", operation.Result);
                return;
            }
        }
    }

    async Task<string> RunOperation(int foo, CancellationToken cancel) {
        log($"[RunOperation] Delaying for: {foo}");
        try {
            await Task.Delay(foo, cancel);
        } catch (OperationCanceledException) {
            log($"[RunOperation] Cancelled");
            throw;
        }
        log($"[RunOperation] Finished");
        return foo.ToString();
    }
}

public static class AggregateExceptionExtensions {
    public static Exception Unwrap(this AggregateException @this) {
        return @this.InnerExceptions.Count == 1 ? @this.InnerExceptions[0] : @this;
    }
}

File: rust/src/internal.rs
use std::mem;
use std::sync::Mutex;
use std::time::Duration;
use futures::{ Future, Poll, Async };
use tokio_core::reactor::{ Handle, Timeout };

use ::MessageBus;

pub struct ExampleMessageBus {
}

impl MessageBus for ExampleMessageBus {
    fn publish(&self, topic: &str, message: String) -> Box<Future<Item=(), Error=()>> {
        unimplemented!()
    }
    fn receive(&self, topic: &str) -> Box<Future<Item=String, Error=()>> {
        unimplemented!()
    }
}

pub struct Service<'a> {
    bus: &'a MessageBus,
    handle: Handle,
    running: Mutex<bool>,
}

impl<'a> Service<'a> {
    pub fn new(bus: &'a MessageBus, handle: Handle) -> Service<'a> {
        Service {
            bus: bus,
            handle: handle,
            running: Mutex::new(false),
        }
    }

    pub fn run<'b>(&'b self) -> Box<Future<Item=(), Error=()> + 'b> {
        *self.running.lock().unwrap() = true;
        return Box::new(Runner {
            service: self,
            state: State::Start,
        });

        enum State {
            Start,
            Receiving { receive: Box<Future<Item=String, Error=()>> },
            Operation { msg: String, receive: Box<Future<Item=String, Error=()>>, op: Timeout },
            Publishing { publish: Box<Future<Item=(), Error=()>> },
            Consumed,
        }

        struct Runner<'a, 'b: 'a> {
            service: &'a Service<'b>,
            state: State,
        }

        impl<'a, 'b> Future for Runner<'a, 'b> {
            type Item = ();
            type Error = ();

            fn poll(&mut self) -> Poll<(), ()> {
                match mem::replace(&mut self.state, State::Consumed) {
                    State::Start => {
                        if *self.service.running.lock().unwrap() {
                            self.state = State::Receiving {
                                receive: self.service.bus.receive("/service"),
                            };
                            self.poll()
                        } else {
                            Ok(Async::Ready(()))
                        }
                    }

                    State::Receiving { mut receive } => {
                        match try!(receive.poll()) {
                            Async::NotReady => {
                                self.state = State::Receiving {
                                    receive: receive
                                };
                                Ok(Async::NotReady)
                            }
                            Async::Ready(msg) => {
                                // log($"[Run] Got start message: {msg}");
                                let foo = Duration::from_millis(msg.parse().unwrap());
                                // log($"[RunOperation] Delaying for: {foo}");
                                self.state = State::Operation {
                                    msg: msg,
                                    receive: self.service.bus.receive("/service"),
                                    op: Timeout::new(foo, &self.service.handle).unwrap(),
                                };
                                self.poll()
                            }
                        }
                    }

                    State::Operation { msg, mut receive, mut op } => {
                        match try!(receive.poll()) {
                            Async::NotReady => {
                                match try!(op.poll().map_err(|_| ())) {
                                    Async::NotReady => {
                                        self.state = State::Operation {
                                            msg: msg,
                                            receive: receive,
                                            op: op,
                                        };
                                        Ok(Async::NotReady)
                                    }
                                    Async::Ready(()) => {
                                        // log($"[RunOperation] Finished");
                                        self.state = State::Publishing {
                                            publish: self.service.bus.publish("/service/op", msg),
                                        };
                                        self.poll()
                                    }
                                }
                            }
                            Async::Ready(msg) => {
                                self.state = if msg == "cancel" {
                                    State::Start
                                } else {
                                    State::Operation {
                                        msg: msg,
                                        receive: receive,
                                        op: op,
                                    }
                                };
                                self.poll()
                            }
                        }
                    }

                    State::Publishing { mut publish } => {
                        match try!(publish.poll()) {
                            Async::NotReady => {
                                self.state = State::Publishing {
                                    publish: publish
                                };
                                Ok(Async::NotReady)
                            }
                            Async::Ready(()) => {
                                self.state = State::Start;
                                self.poll()
                            }
                        }
                    }

                    State::Consumed => {
                        panic!("Already consumed")
                    }
                }
            }
        }
    }

    pub fn stop(&self) {
        *self.running.lock().unwrap() = false;
    }
}
